src/server/util/device-type.ts

```diff
@@ -0,0 +1,26 @@
+import { UAParser } from 'ua-parser-js'
+
+/**
+ * Infers the type of device from a user agent string.
+ *
+ * @param userAgent The user agent string to infer device type from.
+ */
+export function getDeviceType(userAgent: string) {
+  const parser = new UAParser(userAgent)
+  const deviceType = parser.getDevice().type
+
+  // Desktop browsers do not identify themselves in user agent.
+  if (!deviceType) return 'desktop'
```

Comment:
can we consider whether the OS is able to be detected before determining whether desktop is acceptable, as per [this thread](https://github.com/faisalman/ua-parser-js/issues/182)?

Reply:
I took some time to explore this, and found it very easy to extract the OS using the [useragent](https://www.npmjs.com/package/useragent) library. But doing this will mean that we have to keep our list of desktop OS updated. This can also be especially hard with linux distributions uniquely identifying themselves. Do you think we should really do this?

Reply:
my concern is that all bot traffic will be classified as "desktop" since they are unlikely to have a device type. also UAParser can help via `getOS()`, right?

Reply:
As discussed, our check for bots will be: first to check for a JavaScript engine, followed by matching keywords found in common bots